cibuildwheel/linux.py

```diff
@@ -111,201 +102,133 @@ def build(options: BuildOptions) -> None:
         raise Exception('package_dir must be inside the working directory')
 
     container_package_dir = PurePath('/project') / abs_package_dir.relative_to(cwd)
+    container_output_dir = PurePath('/output')
 
     for implementation, platform_tag, docker_image in platforms:
         platform_configs = [c for c in python_configurations if c.identifier.startswith(implementation) and c.identifier.endswith(platform_tag)]
         if not platform_configs:
             continue
 
-        shell_cmd = ['linux32', '/bin/bash'] if platform_tag.endswith("i686") else ['/bin/bash']
+        try:
+            with DockerContainer(docker_image, simulate_32_bit=platform_tag.endswith('i686')) as docker:
+                docker.copy_into(Path.cwd(), Path('/project'))
+
+                if options.before_all:
+                    env = docker.get_environment()
+                    env['PATH'] = f'/opt/python/cp38-cp38:{env["PATH"]}'
+                    env = options.environment.as_dictionary(env, executor=docker.environment_executor)
+
+                    before_all_prepared = prepare_command(options.before_all, project='/project', package=container_package_dir)
+                    docker.call(['sh', '-c', before_all_prepared], env=env)
+
+                for config in platform_configs:
+                    dependency_constraint_flags: List[Union[str, PathLike]] = []
```

Comment:
Unrelated to this PR, I think: If you put this as `else`-clause, you won't need the annotation (since the first occurrence would have the right type).

Reply:
The first occurrence resolves to `List[object]` unfortunately, so needs to be explicit